lib/spack/spack/build_systems/cmake.py

```diff
@@ -375,6 +382,39 @@ def std_args(pkg, generator=None):
             ]
         )
 
+        # Establish some CMake based sandboxing from the user env if package allows it
+        # some packages are incompatible due to their handling of find_package w.r.t. their
+        # dependencies (i.e. some packages vendor find modules to which their build system is hard coded)
+        # and switching to using a config module vendored by a dependnecy will result in dropped libs or includes
+        # from link and compile lines
+        if pkg.enable_cmake_sandboxing:
+            args.extend(
+                [
+                    # CMake bundles a number of find modules with
+                    # its distribution. During package search these
+                    # find modules are prefered in the search order.
+                    # Spack based installations of CMake pacakges
+                    # export their interface in the form of config modules
+                    # by default these are secondary to the find modules
+                    # Ensure CMake prefers Spack derived modules always
+                    # by prefering config modules
+                    define("CMAKE_FIND_PACKAGE_PREFER_CONFIG", True),
```

Comment:
This does not sound like a good idea to me... There are cases where FindZLIB.cmake should be used because Spack built zlib does not have an associated config file, whereas it may exist in a directory of an external.

I don't see what this define has to do with "sandboxing", so better to drop it.

Reply:
I addressed this _exact_ concern in comments in multiple places in this PR. Please read them.

> There are cases where FindZLIB.cmake should be used because Spack built zlib does not have an associated config file, whereas it may exist in a directory of an external.

This is addressed inherently by the use of this CMake argument. The _`PREFER`_ section of the argument name correctly implies that this will only _prefer_ config modules, it will not use them exclusively. It simply inverts the order of the search to look for config modules first, and failing to find them, falls back on find modules rather than the opposite default behavior.

> so the better default is FindZLIB.cmake

The better default is whatever relies exclusively on Spack derived contents, vendored find modules can do the wrong thing, this exact case w.r.t zlib is what motivated this PR. This PR does not preclude the use of the find modules, it prefers the configs.

> The build system may also expect it to use FindZLIB.cmake which exports different variables than a ZLIB config file (iirc).

This is a very valid concern, and one of which I thought of and addressed by means of a per pacakge opt out of this sandboxing. Initially I was adding patches/upstream submissions to fix these packages, but that grew infeasible pretty quickly. I realize the opt out is cludge-y, so if you have constructive thoughts, I'd be happy to hear them.

> define has to do with sandboxing

It's "sandboxing" in the sense that we're prefering configs (which are generally derived directly from Spack packages) and do not bring in anything but Spack based information over vendored find modules which do not originate from Spack based behavior and can accidentally bring in system info.

> so better to submit it as a separate PR if you want this change?

this feels arbitrary based on a very specific and narrow view of sandboxing

Reply:
I know what prefer means :) I also think I'm right in that this config option is orthogonal to sandboxing.

Consider this where it makes things only worse:

Try `fedora:40` with a Spack installed zlib mounted, run `dnf install zlib-devel`, create the following `CMakeLists.txt` file:

```cmake
cmake_minimum_required(VERSION 3.20)
project(Example C)
find_package(ZLIB REQUIRED)
```

and run:

```
cmake . --debug-find -DCMAKE_PREFIX_PATH=/path/to/spack/linux-ubuntu24.04-zen2/gcc-13.2.0/zlib-ng-2.2.1-2uvj6hw3ioeq5x6phso6gcxgbqcqpomo -DCMAKE_FIND_PACKAGE_PREFER_CONFIG=ON
```

it finds system zlib.

Run:

```
cmake . --debug-find -DCMAKE_PREFIX_PATH=/path/to/spack/linux-ubuntu24.04-zen2/gcc-13.2.0/zlib-ng-2.2.1-2uvj6hw3ioeq5x6phso6gcxgbqcqpomo
```

it finds Spack zlib.

Would you consider every package that depends on zlib to opt out of sandboxing then? :)